### PR TITLE
Fix weight decay exclusions in `run_*_no‑trainer.py` examples

### LINKED ISSUE #42754: Excluding weight decay not working properly on most LMs
The problem is consistent across all three files:
https://github.com/huggingface/transformers/blob/471d7ce9abbb3bc1b3bab673367378f9dbc3caac/examples/pytorch/language-modeling/run_clm_no_trainer.py#L518-L528
https://github.com/huggingface/transformers/blob/471d7ce9abbb3bc1b3bab673367378f9dbc3caac/examples/pytorch/language-modeling/run_fim_no_trainer.py#L710-L720
https://github.com/huggingface/transformers/blob/471d7ce9abbb3bc1b3bab673367378f9dbc3caac/examples/pytorch/language-modeling/run_mlm_no_trainer.py#L544-L554
Many LMs use patterns like `[r"layernorm", r"rmsnorm", r"(?:^|\.|_)norm(?:$|\.|_)", r"(?:^|\.|_)ln(?:$|\.|_)"]`, so the simple `no_decay` patterns here won't exclude many normalization layers. I also suggest excluding embeddings and `r"(?:^|\.)wpe(?:$|\.)"` to exclude input/output/positional embeddings.
Could consider using/modifying from the code here:
https://github.com/huggingface/transformers/blob/471d7ce9abbb3bc1b3bab673367378f9dbc3caac/src/transformers/trainer.py#L1213-L1227

### PR DESCRIPTION
**Pull request by casinca**
# What does this PR do?

fixes https://github.com/huggingface/transformers/issues/42754

I'm re-using the more robust logic from `trainer.py` for the `run_*_no_trainer.py` files

https://github.com/huggingface/transformers/blob/471d7ce9abbb3bc1b3bab673367378f9dbc3caac/src/transformers/trainer.py#L1199C1-L1201C32

There are like 10 others which would require this change,  some also had capital `LayerNorm.weight` (vs `"layer_norm.weight"`) in `no_decay = ["bias", "LayerNorm.weight"]`

But before propagating to other `run_*_no_trainer.py` I prefer to make sure my changes are acceptable.

@wwt17 also mentioned adding to the list embeddings, I can add `nn.Embedding` too but this is up to 🤗 to decide.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

 I think @zucchini-nlp and @SunMarc (for trainer) may be interested.

### CODE REVIEW

**Comment by casinca** [comment]
I forgot about this PR and the issue got closed for inactivity but it still seems relevant.

@Rocketknight1 👋
I think at the time you guys were busy with 5.0 and this PR was overlooked.  
Could you take a look and see if it sounds good to you? or could you redirect me to the appropriate reviewer? I'm not sure my initial tagging was adequate. Thanks.

**Comment by Rocketknight1** [comment]
@bot /style

**Review by Rocketknight1 — APPROVED** [review]
Yes, this looks good, and sorry for the delay! The change is definitely an improvement, and I'm fine with doing something similar for the other `no_trainer` files if you want. We should probably just merge this PR first, though - is there anything else you want to add before I do?

**Comment by casinca** [comment]
> Yes, this looks good, and sorry for the delay! The change is definitely an improvement, and I'm fine with doing something similar for the other `no_trainer` files if you want. We should probably just merge this PR first, though - is there anything else you want to add before I do?

Thanks for coming back to this. I'm fine as it is, I will open a separate PR for the other ones.

**Review by Rocketknight1 — APPROVED** [review]
